google/cloud/client.py

```diff
@@ -188,26 +190,50 @@ class _ClientProjectMixin(object):
     """Mixin to allow setting the project on the client.
 
     :type project: str
-    :param project: the project which the client acts on behalf of. If not
-                    passed falls back to the default inferred from the
-                    environment.
+    :param project:
+        (Optional) the project which the client acts on behalf of. If not
+        passed, falls back to the default inferred from the environment.
+
+    :type credentials: :class:`google.auth.credentials.Credentials`
+    :param credentials:
+        (Optional) credentials used to discover a project, if not passed.
 
     :raises: :class:`EnvironmentError` if the project is neither passed in nor
-             set in the environment. :class:`ValueError` if the project value
-             is invalid.
+             set on the credentials or in the environment. :class:`ValueError`
+             if the project value is invalid.
     """
 
-    def __init__(self, project=None):
-        project = self._determine_default(project)
+    def __init__(self, project=None, credentials=None):
+        # This test duplicates the one from `google.auth.default`, but earlier,
+        # for backward compatibility:  we want the environment variable to
+        # override any project set on the credentials.  See:
+        # https://github.com/googleapis/python-cloud-core/issues/27
+        if project is None:
+            project = os.getenv(
+                environment_vars.PROJECT,
+                os.getenv(environment_vars.LEGACY_PROJECT),
+            )
+
+        # Project set on explicit credentials overrides discovery from
+        # SDK / GAE / GCE.
+        if project is None and credentials is not None:
+            project = getattr(credentials, "project_id", None)
+
+        if project is None:
+            project = self._determine_default(project)
+
```

Comment:
I think `google.auth.default()` already treats the environment project id with highest priority. In that case would it be sufficient to just have this bit of code?

https://github.com/googleapis/google-auth-library-python/blob/37141e4dffc2ba3f3f57c5914544fb8b9cf7d017/google/auth/_default.py#L296-L298

Reply:
@busunkim96 The intent of duplicating the environment check above is to ensure that a project ID set in the environment overrides the one from the credentials (for backward compatibility).

Only if neither one is set to we kick down to the "platform checks" in `_determine_default` (gcloud SDK, GAE, GCE, etc.), which just calls into `google.auth.default`.

Reply:
Ah I see. This seems fine for now. It looks like we unfortunately don't have a spec on how project ID detection is supposed to work in the libraries. 😢